pkg/services/folder/registry.go

```diff
@@ -7,7 +7,7 @@ import (
 )
 
 type RegistryService interface {
-	DeleteInFolder(ctx context.Context, orgID int64, folderUID string) error
+	DeleteInFolder(ctx context.Context, orgID int64, folderUID string, user *user.SignedInUser) error
```

Comment:
The user is required during library panel deletion just for checking the permissions:
grafana/grafana#50241
So I wonder if we could perform this check earlier (in the API or service handler) so that we no need this signature change.
WDYT?

Reply:
This should be doable if we perform the check in the service handler.

So we'd have to pass in the folder id as an extra parameter to `DeleteLibraryElementsInFolder`
```
func (l *LibraryElementService) DeleteLibraryElementsInFolder(c context.Context, signedInUser *user.SignedInUser, folderUID string, folderID int64) error {
	if err := l.RequireEditPermissionsOnFolder(c, signedInUser, folderID); err != nil {
		return err
	}
	return l.deleteLibraryElementsInFolderUID(c, signedInUser, folderUID) //
}
```

This means we'd have to capture `id, err := strconv.ParseInt(web.Params(c.Req)[":id"], 10, 64)` in the [API](https://github.com/grafana/grafana/blob/main/pkg/api/folder.go#L284). Not sure if this field is already being set by the frontend or not. If not, we'd have to start setting it. Is that what you had in mind?

Reply:
in folder service `Delete()` implementation we check for delete permission in folder:
https://github.com/grafana/grafana/blob/e3787de470b5637d3407185040b8cefc5023b040/pkg/services/folder/folderimpl/folder.go#L486-L491

apparently, the request will eventually fail here:
https://github.com/grafana/grafana/blob/e3787de470b5637d3407185040b8cefc5023b040/pkg/services/folder/folderimpl/folder.go#L530-L532
if the used does not have **edit** permission to the folder during deleting the library elements:
https://github.com/grafana/grafana/blob/e3787de470b5637d3407185040b8cefc5023b040/pkg/services/libraryelements/database.go#L757-L759

so what I would rather suggest is to check for both **delete** and **edit** permissions in the folder service and remove the check from the library elements

but that's only an idea: no need to block this work for this

Reply:
Thanks for the input. That makes sense. The issue I see is that the function which checks the edit permissions is also used (indirectly) by `DeleteFolder`:  https://github.com/grafana/grafana/blob/main/pkg/api/folder.go#L277  

`DeleteLibraryElementsInFolder` is the one being called there and it is also the one called by `DeleteInFolder`. It’s a wrapper for `deleteLibraryElementsInFolderUID` where the edit permissions are actually being checked.  

If we remove the edit permissions from `deleteLibraryElementsInFolderUID` and add them in the folder service, we need to also add them to `DeleteFolder`. This is because the `DeleteLibraryElementsInFolder` call takes place before the `hs.LibraryElementService.DeleteLibraryElementsInFolder` takes place before the one for `hs.folderService.Delete`.

  It’s not ideal by my suggestion would be to add a todo note. Once permission questions are resolved and we are ready to start deleting alert rules and library panels regardless of ForceDeleteRules and legacy deletion, then we can remove the call to `hs.LibraryElementService.DeleteLibraryElementsInFolder` from `DeleteFolder` and update the `DeleteInFolder` signature so it doesn’t include signedInUser anymore.

Reply:
Talking to Serge, it sounds like he would be in agreement with proceeding with the current approach and pinging AuthNZ about https://github.com/grafana/grafana-enterprise/issues/5144. We depend on that ticket to eventually simplify the code as described in the todo comment I added in [d4101a6](https://github.com/grafana/grafana/pull/69149/commits/d4101a6f727bbe05cb445087a2acbd178bafbe2b).